src/include/btree_inline.h

```diff
@@ -1643,6 +1658,21 @@ __wt_page_del_visible(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del, bool
     if (page_del == NULL)
         return (true);
 
+    /*
+     * Check whether truncate transaction is committed or not. In case of prepared transactions,
+     * truncate operation is committed in two phases. In the first phase prepared state is set to
+     * resolved and in the later phase the committed flag of all the fast truncated pages is set to
+     * true. The two phase design is to handle any page split cases, when a fast truncated page is
+     * read back and later that page splits, in which case all the new split pages will not have
+     * page_del info.
+     *
+     * There is a window between setting the prepared state to resolved and setting the committed
+     * flag to true. Hence if committed flag is not set, consider it as not committed irrespective
+     * of visibility check, to avoid checking visibility when transaction commit is in progress.
+     */
+    if (!__wt_page_del_committed(session, page_del))
```

Comment:
Nice catch about visibility in our own transaction. I don't think your proposed fix is quite right - the `wt_page_del_committed` *should* be a true or false answer. Even if the reason it's uncommitted is because it belongs to the transaction in the current session.

I also think that `visible_all` checks should not include the current transaction when deciding on the result.

So.. I think the logic in this `page_del_visible` function needs to become more complex - i.e: This would become something like:

```
if (!__wt_page_del_committed(session, page_del) &&
    !(F_ISSET(session->txn, WT_TXN_HAS_ID) && id == session->txn->id)) {
    return (false);
}
```

There could be a better way to structure that, but I thought an example might help explain what I meant.

Reply:
Gonna chime in here and say that this `should be a true or false answer. Even if the reason it's uncommitted is because it belongs to the transaction in the current session.` kinda worries me? I think this function has already caused an issue as demonstrated by this ticket. Should we make the function name more verbose? Like `__wt_page_del_committed_set`.

Reply:
Thanks - I think this makes sense, but after applying this change I'm now seeing cache stuck issues again. This makes me think the issue might not be as simple as "check that our txn id matches", but I'm not sure. I'll try to fit in a little more debugging today.

Reply:
We could change the name if you think it could be better.
I think it's reasonable at the moment. It checks whether a deleted page has been marked as committed. That is what most callers need.
This new caller needed something subtly different - since it wanted to know if the delete was committed OR the current transaction did the delete. That is essentially a snapshot isolation visible check. I'm actually kind of concerned about how this layers in with a "regular" visibility check which handles things like read-committed and read-uncommitted isolation levels, but it's probably OK with the simpler change already described.

Reply:
My preference would then be `__wt_page_del_committed_set` as that indicates its checking the state of a flag and not determining whether a thing is `committed` or not.

Reply:
After doing some testing with this and reading code, I don't believe it's correct to only check that `id == session->txn->id` - for example, if the dhandle is a checkpoint, then we may need to look at `session->txn->checkpoint_read_timestamp` instead. (I'm calling this case out since it seems to have been the cause of the test failure, but there are probably other cases.)

Given this, I'd prefer to:
1. Leave `__wt_page_del_committed` as-is, possibly renaming it as Luke suggested.
2. Have the two callers in this file call `__wt_txn_visible` themselves.

Does anyone have feedback (positive or negative) on this?

Reply:
In part 1. would you be reverting the current changes? I think your suggested solution is better as a full visibility check prevents issues like what you saw with the checkpoint read timestamp. I also would prefer to go with the function rename.

Reply:
> In part 1. would you be reverting the current changes?

Yes, sorry that was unclear.

> I also would prefer to go with the function rename.

Ack, I'll rename it then. That part seems fairly uncontroversial. I still want to hear back from @agorrod to see if there's something we're missing before I go ahead with the full change though.

Reply:
Further to today's discussion - I've changed the code in `__wt_page_del_visible_all` to an assertion, thankfully `__wt_txn_visible_all` does seem to do the right thing here.

`__wt_page_del_visible` has been refactored a bit, I moved the existing assertion up so we're checking it as often as we were before, and we immediately bail if `__wt_txn_visible` is false. If it's true, we still need to check the committed flag and whether we're hiding prepared updates, but if those checks don't trip it's visible.

Reply:
Minor update: the assertion tripped so I've made it back into a check and made WT-10477.